pallets/parachain-staking/src/migrations.rs

```diff
@@ -0,0 +1,182 @@
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Moonbeam.  If not, see <http://www.gnu.org/licenses/>.
+
+//! # Migrations
+use crate::{
+	pallet::{migrate_nominator_to_delegator_state, RoundIndex},
+	BalanceOf, CandidateState, CollatorCandidate, CollatorState2, Config, DelegatorState,
+	ExitQueue2, NominatorState2,
+};
+use frame_support::{
+	pallet_prelude::PhantomData,
+	traits::{Get, OnRuntimeUpgrade},
+	weights::Weight,
+};
+use sp_std::collections::btree_map::BTreeMap;
+
+/// Migration to replace the automatic ExitQueue with a manual exits API.
+/// This migration is idempotent so it can be run more than once without any risk.
+/// Returns (reads, writes)
+pub struct RemoveExitQueue<T>(PhantomData<T>);
+impl<T: Config> OnRuntimeUpgrade for RemoveExitQueue<T> {
+	fn on_runtime_upgrade() -> Weight {
+		log::info!(target: "RemoveExitQueue", "running migration to remove staking exit queue");
+		let exit_queue = <ExitQueue2<T>>::take();
+		let (mut reads, mut writes) = (1u64, 0u64);
+		let mut delegator_exits: BTreeMap<T::AccountId, RoundIndex> = BTreeMap::new();
+		let mut delegation_revocations: BTreeMap<T::AccountId, (T::AccountId, RoundIndex)> =
+			BTreeMap::new();
+		// Track scheduled delegator exits and revocations before migrating state
+		// Candidates already track exit info locally so no tracking is necessary
+		for (delegator, is_revocation, when) in exit_queue.nominator_schedule {
+			if let Some(revoking_candidate) = is_revocation {
+				delegation_revocations.insert(delegator, (revoking_candidate, when));
+			} else {
+				delegator_exits.insert(delegator, when);
+			}
+		}
+		// execute candidate migration
+		for (candidate_id, collator_state) in <CollatorState2<T>>::drain() {
+			let candidate_state: CollatorCandidate<T::AccountId, BalanceOf<T>> =
+				collator_state.into();
+			<CandidateState<T>>::insert(candidate_id, candidate_state);
+			reads += 1u64;
+			writes += 1u64;
+		}
+		// execute delegator migration
+		for (delegator_id, nominator_state) in <NominatorState2<T>>::drain() {
+			let mut delegator_state =
+				migrate_nominator_to_delegator_state::<T>(delegator_id.clone(), nominator_state);
+			// add exit if it exists
+			if let Some(when) = delegator_exits.get(&delegator_id) {
+				delegator_state.set_leaving(*when);
+			}
+			// add revocation if exists
+			if let Some((candidate, when)) = delegation_revocations.get(&delegator_id) {
+				delegator_state.hotfix_set_revoke::<T>(candidate.clone(), *when);
+			}
+			<DelegatorState<T>>::insert(delegator_id, delegator_state);
+			reads += 1u64;
+			writes += 1u64;
+		}
+		let db_weight = T::DbWeight::get();
+		if reads > 1u64 {
+			// 10% of the max block weight as safety margin for computation
+			db_weight.reads(reads) + db_weight.writes(writes) + 50_000_000_000
+		} else {
+			// migration was already executed before
+			db_weight.reads(reads)
+		}
```

Comment:
The else clause is when the migration has already executed. Even if it has already executed, it has a `take()` at the beginning of the migration so 1 is the minimum number of reads in this branch (and there are no writes).

Related question, does `take` cost a read and write or just a read? cc @notlesh

Reply:
Removing a storage item definitely incurs I/O cost if there is actually something to remove. It may be a no-op if there is nothing to remove, but it's definitely safer to assume it will cost a write.

What I'm less clear on is what happens when you remove multiple items at once (e.g. remove an entire storage map).